src/binwalk/magic/firmware

```diff
@@ -271,7 +271,7 @@
 #
 # Motorola S-Records, from Gerd Truschinski <gt@freebsd.first.gmd.de>
 # Improved by Martin Sundhaug <martinsundhaug@gmail.com>
-0   regex       (S[0-35-9]([0-9A-F]{4})([0-9A-F]{2})+\n)+       Motorola S-Record{many}; binary data in text format, record type:
```

Comment:
I think you mis-understand: binwalk reads in 1MB of data at a time and applies the signatures across each 1MB block separately. So if you have an S-record or Intel hex file that is, say, 2MB in size, you'll get multiple results even though they are all part of the same input file. The `{many}` tag prevents this. 

Here are the results of running binwalk on a large S-record file without the `{many}` tag:

```
$ binwalk -m srecord_no_many_tag.magic foo.hex 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Motorola S-Record; binary data in text format, record type: header
1048644       0x100044        Motorola S-Record; binary data in text format, record type: data (24-bit)
2097153       0x200001        Motorola S-Record; binary data in text format, record type: data (24-bit)
3145739       0x30000B        Motorola S-Record; binary data in text format, record type: data (24-bit)
4194325       0x400015        Motorola S-Record; binary data in text format, record type: data (24-bit)
5242911       0x50001F        Motorola S-Record; binary data in text format, record type: data (24-bit)
6291497       0x600029        Motorola S-Record; binary data in text format, record type: data (24-bit)
7340083       0x700033        Motorola S-Record; binary data in text format, record type: data (24-bit)
8388669       0x80003D        Motorola S-Record; binary data in text format, record type: data (24-bit)
9437255       0x900047        Motorola S-Record; binary data in text format, record type: data (24-bit)
10485764      0xA00004        Motorola S-Record; binary data in text format, record type: data (24-bit)
11534350      0xB0000E        Motorola S-Record; binary data in text format, record type: data (24-bit)
12582936      0xC00018        Motorola S-Record; binary data in text format, record type: data (24-bit)
13631522      0xD00022        Motorola S-Record; binary data in text format, record type: data (24-bit)
14680108      0xE0002C        Motorola S-Record; binary data in text format, record type: data (24-bit)
15728694      0xF00036        Motorola S-Record; binary data in text format, record type: data (24-bit)
16777280      0x1000040       Motorola S-Record; binary data in text format, record type: data (24-bit)
17825866      0x110004A       Motorola S-Record; binary data in text format, record type: data (24-bit)
18874375      0x1200007       Motorola S-Record; binary data in text format, record type: data (24-bit)
19922961      0x1300011       Motorola S-Record; binary data in text format, record type: data (24-bit)
20971547      0x140001B       Motorola S-Record; binary data in text format, record type: data (24-bit)
22020133      0x1500025       Motorola S-Record; binary data in text format, record type: data (24-bit)
23068719      0x160002F       Motorola S-Record; binary data in text format, record type: data (24-bit)
24117305      0x1700039       Motorola S-Record; binary data in text format, record type: data (24-bit)
25165891      0x1800043       Motorola S-Record; binary data in text format, record type: data (24-bit)
26214400      0x1900000       Motorola S-Record; binary data in text format, record type: data (24-bit)
27262986      0x1A0000A       Motorola S-Record; binary data in text format, record type: data (24-bit)
28311572      0x1B00014       Motorola S-Record; binary data in text format, record type: data (24-bit)
29360158      0x1C0001E       Motorola S-Record; binary data in text format, record type: data (24-bit)
30408744      0x1D00028       Motorola S-Record; binary data in text format, record type: data (24-bit)
31457330      0x1E00032       Motorola S-Record; binary data in text format, record type: data (24-bit)
32505916      0x1F0003C       Motorola S-Record; binary data in text format, record type: data (24-bit)
33554496      0x2000040       Motorola S-Record; binary data in text format, record type: data (24-bit)
34603082      0x210004A       Motorola S-Record; binary data in text format, record type: data (24-bit)
35651591      0x2200007       Motorola S-Record; binary data in text format, record type: data (24-bit)
36700177      0x2300011       Motorola S-Record; binary data in text format, record type: data (24-bit)
37748763      0x240001B       Motorola S-Record; binary data in text format, record type: data (24-bit)
38797349      0x2500025       Motorola S-Record; binary data in text format, record type: data (24-bit)
39845935      0x260002F       Motorola S-Record; binary data in text format, record type: data (24-bit)
40894521      0x2700039       Motorola S-Record; binary data in text format, record type: data (24-bit)
41943107      0x2800043       Motorola S-Record; binary data in text format, record type: data (24-bit)
42991616      0x2900000       Motorola S-Record; binary data in text format, record type: data (24-bit)
44040202      0x2A0000A       Motorola S-Record; binary data in text format, record type: data (24-bit)
45088788      0x2B00014       Motorola S-Record; binary data in text format, record type: data (24-bit)
46137374      0x2C0001E       Motorola S-Record; binary data in text format, record type: data (24-bit)
47185960      0x2D00028       Motorola S-Record; binary data in text format, record type: data (24-bit)
48234546      0x2E00032       Motorola S-Record; binary data in text format, record type: data (24-bit)
49283132      0x2F0003C       Motorola S-Record; binary data in text format, record type: data (24-bit)
50331718      0x3000046       Motorola S-Record; binary data in text format, record type: data (24-bit)
51380227      0x3100003       Motorola S-Record; binary data in text format, record type: data (24-bit)
52428813      0x320000D       Motorola S-Record; binary data in text format, record type: data (24-bit)
53477399      0x3300017       Motorola S-Record; binary data in text format, record type: data (24-bit)
54525985      0x3400021       Motorola S-Record; binary data in text format, record type: data (24-bit)
55574571      0x350002B       Motorola S-Record; binary data in text format, record type: data (24-bit)
56623157      0x3600035       Motorola S-Record; binary data in text format, record type: data (24-bit)
57671743      0x370003F       Motorola S-Record; binary data in text format, record type: data (24-bit)
58720329      0x3800049       Motorola S-Record; binary data in text format, record type: data (24-bit)
59768838      0x3900006       Motorola S-Record; binary data in text format, record type: data (24-bit)
60817424      0x3A00010       Motorola S-Record; binary data in text format, record type: data (24-bit)
61866010      0x3B0001A       Motorola S-Record; binary data in text format, record type: data (24-bit)
62914596      0x3C00024       Motorola S-Record; binary data in text format, record type: data (24-bit)
63963182      0x3D0002E       Motorola S-Record; binary data in text format, record type: data (24-bit)
65011768      0x3E00038       Motorola S-Record; binary data in text format, record type: data (24-bit)
```
And here is a scan of the same file, this time with the `{many}` tag:

```
$ binwalk -m srecord_with_many_tag.magic foo.hex 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Motorola S-Record; binary data in text format, record type: header
```

As you pointed out, in the event that you have multiple S-record files concatenated together, this can unfortunately prevent identification of multiple S-record files. There are, as I see it, three scenarios here:

1. Multiple S-record files are concatenated back-to-back. In this case, due to the regex signature, it doesn't matter how binwalk reads in data or applies the signature, this would still be reported as a single S-record.
2. Multiple S-record files are contained in the same file, but with some other data that binwalk signatures on in between them (some gzip data, a file system, whatever). In this case, the `{many}` tag really has no effect on reporting the multiple the S-record files.
3. Multiple S-record files are contained in the same file, but with some data in between them that binwalk does not recognize (maybe a bunch of NULL padding for example). This is the only case that I see where the `{many}` tag really causes problems, because as you pointed out only the first S-record match will be reported.

I'd argue that while all of these situations are unlikely, option 3 - the only one where the `{many}` tag has a negative impact - is the least likely.